src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala

```diff
@@ -146,20 +145,17 @@ private[mock] class NativeMockKafkaProducer(
       keyDeserializer: KeyDeserializer[IO, K],
       valueDeserializer: ValueDeserializer[IO, V]
   ): IO[List[(Int, K, Option[V])]] = mockProducer.history.asScala.zipWithIndex.toList
-    .traverse { case (record, index) =>
-      for {
-        isSelected <- recordSelector(record)
-        key <-
-          if (isSelected) keyDeserializer.deserialize(topic, Headers.empty, record.key).map(Option(_))
-          else IO(None)
-        value <-
-          (if (isSelected) {
-             if (record.value eq null) IO(Some(None))
-             else valueDeserializer.deserialize(topic, Headers.empty, record.value).map(v => Some(Some(v)))
-           } else IO(None)): IO[Option[Option[V]]]
-      } yield key.zip(value).map { case (k, v) => (index, k, v) }
+    .flatTraverse { case (record, index) =>
+      recordSelector(record)
+        .ifM(
```

Comment:
I tried hard to get rid of this if and somehow lift the condition to Some/None or something more structure, but couldn't figure it out.